OptimizelySDK/OptimizelyUserContext.cs

```diff
@@ -0,0 +1,144 @@
+﻿/* 
+ * Copyright 2020, Optimizely
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+using OptimizelySDK.Logger;
+using System.Collections.Generic;
+using OptimizelySDK.ErrorHandler;
+using OptimizelySDK.Entity;
+using OptimizelySDK.OptimizelyDecisions;
+using System;
+
+namespace OptimizelySDK
+{
+    /// <summary>
+    /// OptimizelyUserContext defines user contexts that the SDK will use to make decisions for
+    /// </summary>
+    public class OptimizelyUserContext
+    {
+        private ILogger Logger;
+        private IErrorHandler ErrorHandler;
+        // userID for Optimizely user context
+        public string UserId { get; }
+        // user attributes for Optimizely user context.
+        public UserAttributes UserAttributes { get; }
+        // Optimizely object to be used.
+        public Optimizely Optimizely { get; }
+
+        public OptimizelyUserContext(Optimizely optimizely, string userId, UserAttributes userAttributes, IErrorHandler errorHandler, ILogger logger)
+        {
+            ErrorHandler = errorHandler;
+            Logger = logger;
+            Optimizely = optimizely;
+            UserAttributes = userAttributes ?? new UserAttributes();
+            UserId = userId;
+        }
+
+        /// <summary>
+        /// Set an attribute for a given key.
+        /// </summary>
+        /// <param name="key">An attribute key</param>
+        /// <param name="value">value An attribute value</param>
+        public void SetAttribute(string key, object value)
+        {
+            if (key == null)
+            {
+                Logger.Log(LogLevel.WARN, "Null attribute key.");
+            }
+            else
+            { 
+                UserAttributes.Add(key, value);
```

Comment:
Need a mutex lock - setAttribute and reading attributes can happen concurrently.